File: lib/koto_cms/miauth.ex
defmodule KotoCms.MiAuth do
  alias KotoCms.Allowlist

  def initiate_flow(handle) do
    case validate_handle(handle) do
      {:ok, %{username: _username, instance: instance}} ->
        instance_url = "https://#{instance}"
        session_id = UUID.uuid4()
        callback_url = System.get_env("MIAUTH_CALLBACK_URL") || "http://localhost:3000/miauth/callback"
        app_name = System.get_env("APP_NAME") || "Koto"

        auth_url = "#{instance_url}/miauth/#{session_id}?name=#{URI.encode(app_name)}&callback=#{URI.encode(callback_url)}&permission=read:account"

        {:ok, %{sessionUrl: auth_url, sessionId: session_id}}

      {:error, reason} -> {:error, reason}
    end
  end

  def handle_callback(session_id, instance_url) do
    check_url = "#{instance_url}/api/miauth/#{session_id}/check"

    case Req.post(check_url, json: %{}) do
      {:ok, %{status: 200, body: body}} ->
        process_auth_response(body, instance_url)

      {:ok, %{status: status}} ->
        {:error, "MiAuth check failed: #{status}"}

      {:error, reason} ->
        {:error, "MiAuth error: #{inspect(reason)}"}
    end
  end

  defp process_auth_response(%{"user" => user} = data, instance_url) do
    mi_token = data["token"]
    instance = user["host"] || URI.parse(instance_url).host
    fedi_handle = "@#{user["username"]}@#{instance}"

    case Allowlist.validate(fedi_handle) do
      false ->
        {:error, "Editor #{fedi_handle} is not in the allowlist"}

      true ->
        display_name = fetch_display_name(user, instance_url, mi_token)
        github_token = System.get_env("GITHUB_BOT_TOKEN")

        if github_token do
          {:ok, %{
            ok: true,
            access_token: github_token,
            author_name: display_name,
            author_email: "#{user["username"]}+#{instance}@users.noreply.fediverse",
            fedi_handle: fedi_handle
          }}
        else
          {:error, "Bot GitHub token not configured"}
        end
    end
  end

  defp process_auth_response(_, _) do
    {:error, "User not authenticated"}
  end

  defp fetch_display_name(user, instance_url, mi_token) do
    profile_url = "#{instance_url}/api/users/show"
    body = %{username: user["username"], host: user["host"], i: mi_token}

    case Req.post(profile_url, json: body) do
      {:ok, %{status: 200, body: profile}} ->
        profile["name"] || user["name"] || user["username"]

      _ ->
        user["name"] || user["username"]
    end
  end

  defp validate_handle(handle) do
    normalized = String.replace_prefix(handle, "@", "")

    case String.split(normalized, "@") do
      [username, instance] when username != "" and instance != "" ->
        if String.contains?(instance, ".") or instance == "localhost" do
          {:ok, %{username: username, instance: instance}}
        else
          {:error, "Invalid instance domain"}
        end

      _ ->
        {:error, "Invalid handle format"}
    end
  end
end

File: infrastructure/cloudflare-do/src/miauth.ts
export interface MiauthInitResult {
  sessionUrl: string;
  sessionId: string;
}

export interface MiauthUser {
  username: string;
  host?: string;
  name?: string;
}

export interface MiauthCallbackResult {
  ok: true;
  author_name: string;
  author_email: string;
  fedi_handle: string;
}

export class MiauthError extends Error {}

function validateHandle(handle: string): { username: string; instance: string } {
  const normalized = handle.replace(/^@/, '');
  const [username, instance] = normalized.split('@');
  if (!username || !instance) {
    throw new MiauthError('Invalid handle format');
  }
  if (!instance.includes('.') && instance !== 'localhost') {
    throw new MiauthError('Invalid instance domain');
  }
  return { username, instance };
}

export function initiateFlow(opts: {
  handle: string;
  appName: string;
  callbackUrl: string;
}): MiauthInitResult {
  const { instance } = validateHandle(opts.handle);
  if (!opts.callbackUrl) {
    throw new MiauthError('callbackUrl not configured');
  }
  const sessionId = crypto.randomUUID();
  const params = new URLSearchParams({
    name: opts.appName,
    callback: opts.callbackUrl,
    permission: 'read:account',
  });
  const sessionUrl = `https://${instance}/miauth/${sessionId}?${params.toString()}`;
  return { sessionUrl, sessionId };
}

export function instanceOriginForHandle(handle: string): string {
  const { instance } = validateHandle(handle);
  return `https://${instance}`;
}

export async function handleCallback(
  sessionId: string,
  instanceUrl: string,
): Promise<MiauthCallbackResult> {
  const checkUrl = `${instanceUrl}/api/miauth/${sessionId}/check`;
  const checkRes = await fetch(checkUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: '{}',
  });
  if (!checkRes.ok) {
    throw new MiauthError(`MiAuth check failed: ${checkRes.status}`);
  }
  const data = (await checkRes.json()) as {
    user?: MiauthUser;
    token?: string;
  };
  if (!data.user) throw new MiauthError('User not authenticated');

  const miToken = data.token ?? '';
  const parsedHost = new URL(instanceUrl).host;
  const instance = data.user.host ?? parsedHost;
  const fediHandle = `@${data.user.username}@${instance}`;
  const displayName = await fetchDisplayName(data.user, instanceUrl, miToken);

  return {
    ok: true,
    author_name: displayName,
    author_email: `${data.user.username}+${instance}@users.noreply.fediverse`,
    fedi_handle: fediHandle,
  };
}

async function fetchDisplayName(
  user: MiauthUser,
  instanceUrl: string,
  miToken: string,
): Promise<string> {
  const profileUrl = `${instanceUrl}/api/users/show`;
  try {
    const res = await fetch(profileUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        username: user.username,
        host: user.host ?? null,
        i: miToken,
      }),
    });
    if (!res.ok) return user.name ?? user.username;
    const profile = (await res.json()) as { name?: string };
    return profile.name || user.name || user.username;
  } catch {
    return user.name ?? user.username;
  }
}
